File: java/src/main/java/it/giannibombelli/wsc2026/payment/domain/payment/Payment.java
package it.giannibombelli.wsc2026.payment.domain.payment;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.domain.model.Aggregate;
import it.giannibombelli.wsc2026.common.domain.primitive.ClientReference;
import it.giannibombelli.wsc2026.common.domain.primitive.Description;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.common.domain.primitive.Timestamp;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentResultEvents;
import it.giannibombelli.wsc2026.payment.domain.events.RefundRequested;
import it.giannibombelli.wsc2026.payment.domain.events.RefundResultEvents;
import it.giannibombelli.wsc2026.payment.domain.events.TransactionStarted;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public final class Payment implements Aggregate<PaymentId> {
    private final PaymentId id;
    private final ClientReference clientReference;
    private final Money amount;
    private final Timestamp requestedAt;
    private PaymentStatus status;
    private final List<Transaction> transactions = new ArrayList<>();

    public Payment(PaymentId id, ClientReference clientReference, Money amount, PaymentStatus status, Timestamp requestedAt, List<Transaction> transactions) {
        Require.requireArgument(id, "id");
        Require.requireArgument(clientReference, "clientReference");
        Require.requireArgument(amount, "amount");
        Require.requireArgument(status, "status");
        Require.requireArgument(requestedAt, "requestedAt");
        Require.requireArgument(transactions, "transactions");
        this.id = id;
        this.clientReference = clientReference;
        this.amount = amount;
        this.status = status;
        this.requestedAt = requestedAt;
        this.transactions.addAll(transactions);
    }

    public PaymentId id() {
        return id;
    }

    public ClientReference clientReference() {
        return clientReference;
    }

    public Money amount() {
        return amount;
    }

    public PaymentStatus status() {
        return status;
    }

    public Timestamp requestedAt() {
        return requestedAt;
    }

    public List<Transaction> transactions() {
        return List.copyOf(transactions);
    }

    public static Payment request(PaymentId id, ClientReference clientReference, Money amount, Timestamp requestedAt) {
        Require.requireArgument(id, "id");
        Require.requireArgument(clientReference, "clientReference");
        Require.requireArgument(amount, "amount");
        Require.requireArgument(requestedAt, "requestedAt");

        return new Payment(id, clientReference, amount, PaymentStatus.REQUESTED, requestedAt, List.of());
    }

    public TransactionStarted startTransaction(TransactionId transactionId, Provider provider, ProviderReference providerReference, Money amount, Timestamp startedAt) {
        Require.requireArgument(transactionId, "transactionId");
        Require.requireArgument(provider, "provider");
        Require.requireArgument(amount, "amount");
        Require.requireArgument(startedAt, "startedAt");
        ensureMutable();

        Transaction transaction = Transaction.start(transactionId, provider, providerReference, amount, startedAt);
        transactions.add(transaction);

        this.status = PaymentStatus.PROCESSING;

        return new TransactionStarted(id, provider, transactionId, amount);
    }

    public PaymentResultEvents.PaymentAccepted acceptTransaction(TransactionId transactionId, Timestamp providerCompletedAt) {
        Require.requireArgument(transactionId, "transactionId");
        Require.requireArgument(providerCompletedAt, "providerCompletedAt");
        ensureMutable();

        Transaction transaction = findTransaction(transactionId);
        Transaction accepted = transaction.accept(providerCompletedAt);
        replaceTransaction(transaction, accepted);

        if (!isWithinAcceptanceWindow(providerCompletedAt)) {
            throw new IllegalStateException("transaction outside acceptance window");
        }

        if (allAcceptedAndTimely() && sufficientCoverage()) {
            this.status = PaymentStatus.ACCEPTED;
            return new PaymentResultEvents.PaymentAccepted(id, clientReference, amount);
        }

        return null;
    }

    public PaymentResultEvents.PaymentRejected rejectTransaction(TransactionId transactionId, Description reason) {
        Require.requireArgument(transactionId, "transactionId");
        Require.requireArgument(reason, "reason");
        ensureMutable();

        Transaction transaction = findTransaction(transactionId);
        Transaction rejected = transaction.reject(Timestamp.now());
        replaceTransaction(transaction, rejected);

        this.status = PaymentStatus.REJECTED;
        return new PaymentResultEvents.PaymentRejected(id, clientReference, amount, reason);
    }

    public PaymentResultEvents.PaymentExpired expire() {
        if (status == PaymentStatus.ACCEPTED) throw new IllegalStateException("payment already accepted");
        this.status = PaymentStatus.EXPIRED;
        return new PaymentResultEvents.PaymentExpired(id, clientReference, amount);
    }

    public RefundRequested requestRefund(Money amount) {
        Require.requireArgument(amount, "amount");
        if (this.status != PaymentStatus.ACCEPTED)
            throw new IllegalStateException("only accepted payments can be refunded");
        if (amount.isGreaterThan(this.amount)) {
            throw new IllegalArgumentException("refund amount cannot exceed payment amount");
        }
        return new RefundRequested(id, clientReference, amount);
    }

    public void markTransactionRefunded(TransactionId transactionId) {
        Require.requireArgument(transactionId, "transactionId");
        if (this.status != PaymentStatus.ACCEPTED)
            throw new IllegalStateException("only accepted payments can be refunded");
        Transaction transaction = findTransaction(transactionId);
        Transaction refunded = transaction.refund(Timestamp.now());
        replaceTransaction(transaction, refunded);
    }

    public RefundResultEvents.TransactionRefunded refundTransaction(Money amount) {
        Require.requireArgument(amount, "amount");
        if (this.status != PaymentStatus.ACCEPTED)
            throw new IllegalStateException("only accepted payments can be refunded");
        transactions.stream()
            .filter(t -> t.status() == TransactionStatus.ACCEPTED)
            .forEach(t -> markTransactionRefunded(t.id()));
        this.status = PaymentStatus.REFUNDED;
        return new RefundResultEvents.TransactionRefunded(id, clientReference, amount);
    }

    public RefundResultEvents.TransactionNotRefunded rejectRefund(Provider provider, ProviderReference providerReference, Description reason) {
        Require.requireArgument(provider, "provider");
        Require.requireArgument(providerReference, "providerReference");
        Require.requireArgument(reason, "reason");
        if (this.status != PaymentStatus.ACCEPTED)
            throw new IllegalStateException("only accepted payments can be refunded");
        return new RefundResultEvents.TransactionNotRefunded(id, clientReference, reason);
    }

    private void ensureMutable() {
        if (status == PaymentStatus.ACCEPTED) throw new IllegalStateException("payment already accepted");
        if (status == PaymentStatus.EXPIRED) throw new IllegalStateException("payment expired");
        if (status == PaymentStatus.REFUNDED) throw new IllegalStateException("payment refunded");
    }

    private Transaction findTransaction(TransactionId transactionId) {
        return transactions.stream()
            .filter(t -> t.id().equals(transactionId))
            .findFirst()
            .orElseThrow(() -> new IllegalStateException("transaction not found"));
    }

    private void replaceTransaction(Transaction oldTransaction, Transaction newTransaction) {
        int index = transactions.indexOf(oldTransaction);
        if (index < 0) throw new IllegalStateException("transaction not found");
        transactions.set(index, newTransaction);
    }

    private boolean isWithinAcceptanceWindow(Timestamp providerCompletedAt) {
        if (requestedAt == null) return false;
        Timestamp deadline = requestedAt.plusSeconds(48 * 3600);
        return providerCompletedAt.isAfterOrEqual(requestedAt) && providerCompletedAt.isBeforeOrEqual(deadline);
    }

    private boolean allAcceptedAndTimely() {
        if (requestedAt == null) return false;
        if (transactions.isEmpty()) return false;
        boolean allAccepted = transactions.stream().allMatch(t -> t.status() == TransactionStatus.ACCEPTED);
        boolean allTimely = transactions.stream().allMatch(t -> isWithinAcceptanceWindow(t.completedAt()));
        return allAccepted && allTimely && sufficientCoverage();
    }

    private boolean sufficientCoverage() {
        Money sum = transactions.stream()
            .filter(t -> t.status() == TransactionStatus.ACCEPTED)
            .map(Transaction::amount)
            .reduce(Money.zero(), Money::plus);
        return !sum.isLessThan(amount);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        return Objects.equals(id, ((Payment) o).id);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }

    @Override
    public String toString() {
        return "Payment[" +
            "id=" + id + ", " +
            "clientReference=" + clientReference + ", " +
            "amount=" + amount + ", " +
            "status=" + status +
            ']';
    }
}

File: typescript/src/payment/domain/payment/payment.ts
// Aggregato Payment: gestisce il ciclo di vita di una richiesta di pagamento,
// delle transazioni e dei rimborsi.

import { ClientReference } from '@/common/domain/primitive/clientReference.js';
import { Description } from '@/common/domain/primitive/description.js';
import { Money } from '@/common/domain/primitive/money.js';
import { Timestamp } from '@/common/domain/primitive/timestamp.js';
import type { Aggregate } from '@/common/domain/model/aggregate.js';
import {
  paymentAccepted,
  paymentRejected,
  paymentExpired,
  type PaymentAccepted,
  type PaymentRejected,
  type PaymentExpired,
} from '../events/paymentResultEvents.js';
import { refundRequested, type RefundRequested } from '../events/refundRequested.js';
import {
  transactionRefunded,
  transactionNotRefunded,
  type TransactionRefunded,
  type TransactionNotRefunded,
} from '../events/refundResultEvents.js';
import { transactionStarted, type TransactionStarted } from '../events/transactionStarted.js';
import { PaymentId } from './paymentId.js';
import { PaymentStatus } from './paymentStatus.js';
import { Provider } from './provider.js';
import { ProviderReference } from './providerReference.js';
import { Transaction } from './transaction.js';
import { TransactionId } from './transactionId.js';
import { TransactionNotFoundException } from './transactionNotFoundException.js';
import { TransactionStatus } from './transactionStatus.js';
import { requireArgument } from '@/common/utils/requireArgument.js';


export class Payment implements Aggregate<PaymentId> {
  private readonly _id: PaymentId;
  private readonly _clientReference: ClientReference;
  private readonly _amount: Money;
  private readonly _requestedAt: Timestamp;
  private _status: PaymentStatus;
  private readonly _transactions: Transaction[];

  constructor(
    id: PaymentId,
    clientReference: ClientReference,
    amount: Money,
    status: PaymentStatus,
    requestedAt: Timestamp,
    transactions: Transaction[]
  ) {
    requireArgument(id, 'id');
    requireArgument(clientReference, 'clientReference');
    requireArgument(amount, 'amount');
    requireArgument(status, 'status');
    requireArgument(requestedAt, 'requestedAt');
    requireArgument(transactions, 'transactions');

    this._id = id;
    this._clientReference = clientReference;
    this._amount = amount;
    this._status = status;
    this._requestedAt = requestedAt;
    this._transactions = [...transactions];
  }

  static request(id: PaymentId, clientReference: ClientReference, amount: Money, requestedAt: Timestamp): Payment {
    requireArgument(id, 'id');
    requireArgument(clientReference, 'clientReference');
    requireArgument(amount, 'amount');
    requireArgument(requestedAt, 'requestedAt');

    return new Payment(id, clientReference, amount, PaymentStatus.REQUESTED, requestedAt, []);
  }

  id(): PaymentId {
    return this._id;
  }

  clientReference(): ClientReference {
    return this._clientReference;
  }

  amount(): Money {
    return this._amount;
  }

  status(): PaymentStatus {
    return this._status;
  }

  requestedAt(): Timestamp {
    return this._requestedAt;
  }

  transactions(): Transaction[] {
    return [...this._transactions];
  }

  startTransaction(
    transactionId: TransactionId,
    provider: Provider,
    providerReference: ProviderReference | null,
    amount: Money,
    startedAt: Timestamp
  ): TransactionStarted {
    requireArgument(transactionId, 'transactionId');
    requireArgument(provider, 'provider');
    requireArgument(amount, 'amount');
    requireArgument(startedAt, 'startedAt');
    this.ensureMutable();

    const transaction = Transaction.start(transactionId, provider, providerReference, amount, startedAt);
    this._transactions.push(transaction);
    this._status = PaymentStatus.PROCESSING;

    return transactionStarted(this._id, provider, transactionId, amount);
  }

  acceptTransaction(transactionId: TransactionId, providerCompletedAt: Timestamp): PaymentAccepted | null {
    requireArgument(transactionId, 'transactionId');
    requireArgument(providerCompletedAt, 'providerCompletedAt');
    this.ensureMutable();

    const transaction = this.findTransaction(transactionId);
    const accepted = transaction.accept(providerCompletedAt);
    this.replaceTransaction(transaction, accepted);

    if (!this.isWithinAcceptanceWindow(providerCompletedAt)) {
      throw new Error('transaction outside acceptance window');
    }

    if (this.allAcceptedAndTimely()) {
      this._status = PaymentStatus.ACCEPTED;
      return paymentAccepted(this._id, this._clientReference, this._amount);
    }

    return null;
  }

  rejectTransaction(transactionId: TransactionId, reason: Description): PaymentRejected {
    requireArgument(transactionId, 'transactionId');
    requireArgument(reason, 'reason');
    this.ensureMutable();

    const transaction = this.findTransaction(transactionId);
    const rejected = transaction.reject(Timestamp.now());
    this.replaceTransaction(transaction, rejected);

    this._status = PaymentStatus.REJECTED;
    return paymentRejected(this._id, this._clientReference, this._amount, reason);
  }

  expire(): PaymentExpired {
    if (this._status === PaymentStatus.ACCEPTED) throw new Error('payment already accepted');
    this._status = PaymentStatus.EXPIRED;
    return paymentExpired(this._id, this._clientReference, this._amount);
  }

  requestRefund(amount: Money): RefundRequested {
    requireArgument(amount, 'amount');
    if (this._status !== PaymentStatus.ACCEPTED) throw new Error('only accepted payments can be refunded');
    if (amount.isGreaterThan(this._amount)) {
      throw new Error('refund amount cannot exceed payment amount');
    }
    return refundRequested(this._id, this._clientReference, amount);
  }

  markTransactionRefunded(transactionId: TransactionId): void {
    requireArgument(transactionId, 'transactionId');
    if (this._status !== PaymentStatus.ACCEPTED) throw new Error('only accepted payments can be refunded');
    const transaction = this.findTransaction(transactionId);
    const refunded = transaction.refund(Timestamp.now());
    this.replaceTransaction(transaction, refunded);
  }

  refundTransaction(amount: Money): TransactionRefunded {
    requireArgument(amount, 'amount');
    if (this._status !== PaymentStatus.ACCEPTED) throw new Error('only accepted payments can be refunded');
    this._transactions
      .filter((t) => t.status() === TransactionStatus.ACCEPTED)
      .forEach((t) => this.markTransactionRefunded(t.id()));
    this._status = PaymentStatus.REFUNDED;
    return transactionRefunded(this._id, this._clientReference, amount);
  }

  rejectRefund(provider: Provider, providerReference: ProviderReference, reason: Description): TransactionNotRefunded {
    requireArgument(provider, 'provider');
    requireArgument(providerReference, 'providerReference');
    requireArgument(reason, 'reason');
    if (this._status !== PaymentStatus.ACCEPTED) throw new Error('only accepted payments can be refunded');
    return transactionNotRefunded(this._id, this._clientReference, reason);
  }

  private ensureMutable(): void {
    if (this._status === PaymentStatus.ACCEPTED) throw new Error('payment already accepted');
    if (this._status === PaymentStatus.EXPIRED) throw new Error('payment expired');
    if (this._status === PaymentStatus.REFUNDED) throw new Error('payment refunded');
  }

  private findTransaction(transactionId: TransactionId): Transaction {
    const transaction = this._transactions.find((t) => t.id().equals(transactionId));
    if (transaction === undefined) throw new TransactionNotFoundException();
    return transaction;
  }

  private replaceTransaction(oldTransaction: Transaction, newTransaction: Transaction): void {
    const index = this._transactions.indexOf(oldTransaction);
    if (index < 0) throw new TransactionNotFoundException();
    this._transactions[index] = newTransaction;
  }

  private isWithinAcceptanceWindow(providerCompletedAt: Timestamp): boolean {
    const deadline = this._requestedAt.plusSeconds(48 * 3600);
    return providerCompletedAt.isAfterOrEqual(this._requestedAt) && providerCompletedAt.isBeforeOrEqual(deadline);
  }

  private allAcceptedAndTimely(): boolean {
    if (this._transactions.length === 0) return false;
    const allAccepted = this._transactions.every((t) => t.status() === TransactionStatus.ACCEPTED);
    const allTimely = this._transactions.every(
      (t) => t.completedAt() !== null && this.isWithinAcceptanceWindow(t.completedAt()!)
    );
    return allAccepted && allTimely && this.sufficientCoverage();
  }

  private sufficientCoverage(): boolean {
    const sum = this._transactions
      .filter((t) => t.status() === TransactionStatus.ACCEPTED)
      .reduce((acc, t) => acc.plus(t.amount()), Money.zero());
    return !sum.isLessThan(this._amount);
  }

  equals(other: Payment): boolean {
    return this._id.value.value === other._id.value.value;
  }

  hashCode(): number {
    return this._id.value.value.split('').reduce((hash, char) => {
      hash = (hash << 5) - hash + char.charCodeAt(0);
      return hash | 0;
    }, 0);
  }
}
